ksql-rest-app/src/main/java/io/confluent/ksql/api/auth/JaasAuthProvider.java

```diff
@@ -0,0 +1,189 @@
+    this.loginContextSupplier =
+        Objects.requireNonNull(loginContextSupplier, "loginContextSupplier");
+  }
+
+  @VisibleForTesting
+  @FunctionalInterface
+  interface LoginContextSupplier {
+    LoginContext get(String name, CallbackHandler callbackHandler) throws LoginException;
+  }
+
+  @Override
+  public void authenticate(
+      final JsonObject authInfo,
+      final Handler<AsyncResult<User>> resultHandler
+  ) {
+    final String username = authInfo.getString("username");
+    if (username == null) {
+      resultHandler.handle(Future.failedFuture("authInfo missing 'username' field"));
+      return;
+    }
+    final String password = authInfo.getString("password");
+    if (password == null) {
+      resultHandler.handle(Future.failedFuture("authInfo missing 'password' field"));
+      return;
+    }
+
+    final String contextName = config.getString(ApiServerConfig.AUTHENTICATION_REALM_CONFIG);
+    final List<String> allowedRoles = config.getList(ApiServerConfig.AUTHENTICATION_ROLES_CONFIG);
+
+    server.getWorkerExecutor().executeBlocking(
+        p -> getUser(contextName, username, password, allowedRoles, p),
+        resultHandler
+    );
+  }
+
+  private void getUser(
+      final String contextName,
+      final String username,
+      final String password,
+      final List<String> allowedRoles,
+      final Promise<User> promise
+  ) {
+    final LoginContext lc;
+    try {
+      lc = loginContextSupplier.get(contextName, new BasicCallbackHandler(username, password));
+    } catch (LoginException | SecurityException e) {
+      log.error("Failed to create LoginContext. " + e.getMessage());
+      promise.fail("Failed to create LoginContext.");
+      return;
+    }
+
+    try {
+      lc.login();
+    } catch (LoginException le) {
+      log.error("Failed to log in. " + le.getMessage());
+      promise.fail("Failed to log in: Invalid username/password.");
+      return;
+    }
+
+    if (!validateRoles(lc, allowedRoles)) {
+      log.error("Failed to log in: Invalid roles.");
+      promise.fail("Failed to log in: Invalid roles.");
+      return;
+    }
+
+    promise.complete(new JaasUser(username, this));
+  }
+
+  private void checkUserPermission(
+      final String username,
+      final Handler<AsyncResult<Boolean>> resultHandler
+  ) {
+    // no authorization yet (besides JAAS role check during login)
+    // consequently, authenticated users have all permissions
+    resultHandler.handle(Future.succeededFuture(true));
+  }
+
+  private static boolean validateRoles(final LoginContext lc, final List<String> allowedRoles) {
+    if (allowedRoles.contains("*")) {
+      // all users allowed
+      return true;
+    }
+
+    final Set<String> userRoles = lc.getSubject().getPrincipals().stream()
+        .map(Principal::getName)
+        .collect(Collectors.toSet());
+    return !CollectionUtils.intersection(userRoles, allowedRoles).isEmpty();
+  }
+
+  @SuppressWarnings("deprecation")
```

Comment:
The `AbstractUser` interface is deprecated in vert.x 3.8.5 along with most methods of the `User` interface, but their replacements (discussed in https://groups.google.com/d/msg/vertx/Nq20qhvUFzo/j5UAE2QCAwAJ) are not available in 3.8.5 so it's not clear to me what should be implemented instead. Thus this suppression.